app/src/main/java/de/westnordost/streetcomplete/util/SpatialCache.kt

```diff
@@ -0,0 +1,179 @@
+package de.westnordost.streetcomplete.util
+
+import android.util.Log
+import de.westnordost.streetcomplete.data.download.tiles.TilePos
+import de.westnordost.streetcomplete.data.download.tiles.enclosingTilePos
+import de.westnordost.streetcomplete.data.download.tiles.enclosingTilesRect
+import de.westnordost.streetcomplete.data.download.tiles.minTileRect
+import de.westnordost.streetcomplete.data.osm.mapdata.BoundingBox
+import de.westnordost.streetcomplete.data.osm.mapdata.LatLon
+import de.westnordost.streetcomplete.util.math.contains
+import de.westnordost.streetcomplete.util.math.isCompletelyInside
+
+/**
+ * Spatial cache containing items of type T that each must have an id of type K and a position
+ * (LatLon). See [getKey] and [getPosition].
+ *
+ * @fetch needs to get data from db and fill other caches, e.g. questKey -> Quest
+ *
+ * the bbox queried in get() must fit in cache, i.e. may not be larger than maxTiles at tileZoom.
+ */
+class SpatialCache<K, T>(
+    private val tileZoom: Int,
+    private val maxTiles: Int,
+    initialCapacity: Int?,
+    private val fetch: (BoundingBox) -> Collection<T>,
+    private val getKey: T.() -> K,
+    private val getPosition: T.() -> LatLon,
+) {
+    private val byTile = LinkedHashMap<TilePos, HashSet<T>>((maxTiles/0.75).toInt(), 0.75f, true)
+    private val byKey = initialCapacity?.let { HashMap<K, T>(it) } ?: HashMap<K, T>()
+    val size get() = byTile.size
+
+    /** @return a new list of all keys in the cache */
+    fun getKeys(): List<K> = synchronized(this) { byKey.keys.toList() }
+
+    /** @return a new list of all tilePos in the cache */
+    fun getTiles(): List<TilePos> = synchronized(this) { byTile.keys.toList() }
```

Comment:
I think there were reasons previously, but now it can definitely be changed.
Same could be done for getKeys, but this is not used at all any more, except in tests.

Reply:
changed `getTiles` to return a set